src/library/scala/collection/StrictOptimizedSeqOps.scala

```diff
@@ -14,15 +14,13 @@ trait StrictOptimizedSeqOps [+A, +CC[_], +C]
   override def distinctBy[B](f: A => B): C = {
     val builder = newSpecificBuilder()
     val seen = mutable.HashSet.empty[B]
-
-    for (x <- this) {
-      val y = f(x)
-      if (!seen.contains(y)) {
-        seen += y
-        builder += x
-      }
+    val it = this.iterator()
+    var different = false
+    while (it.hasNext) {
+      val next = it.next()
+      if (seen.add(f(next))) builder += next else different = true
     }
-    builder.result()
+    if (different) builder.result() else coll
```

Comment:
I remember the same problem from the 2.12.x version. This check is only allowed for immutable collections. Mutable collections must always return a new instance. (But now we can move it down into `immutable.StrictOptimizedSeqOps` instead of adding an `isInstanceOf` check)

Reply:
@szeiger But that is what I did, no? :)

Reply:
@szeiger or if not, then can you elaborate on what changes exactly you would like to see?
Implementation is already spitted between `StrictOptimizedSeqOps.scala` and `StrictOptimizedSeqOps.scala` and there is no `isInstanceOf` here.

Reply:
This is in `scala.collection.StrictOptimizedSeqOps` which is used by both, mutable and immutable collections, but the optimization is only allowed for immutable collections. The implementation in `scala.collection.StrictOptimizedSeqOps` must always build a new collection, on override in `scala.collection.immutable.StrictOptimizedSeqOps` can perform the optimization for immutable collections only.